pageserver/src/walredo.rs

```diff
@@ -682,16 +690,75 @@ impl WalRedoProcess {
         set_nonblock_or_log_err!(stdout)?;
         set_nonblock_or_log_err!(stderr)?;
 
+        let mut stderr = tokio::io::unix::AsyncFd::new(stderr).context("AsyncFd::with_interest")?;
+
         // all fallible operations post-spawn are complete, so get rid of the guard
         let child = scopeguard::ScopeGuard::into_inner(child);
 
+        let stderr_logger_cancel = CancellationToken::new();
+        let (stderr_logger_task_done_tx, stderr_logger_task_done_rx) =
+            tokio::sync::watch::channel(false);
+        tokio::spawn({
+            let stderr_logger_cancel = stderr_logger_cancel.clone();
+            async move {
+                scopeguard::defer! {
+                    debug!("wal-redo-postgres stderr_logger_task finished");
+                    let _ = stderr_logger_task_done_tx.send(true);
+                }
+                debug!("wal-redo-postgres stderr_logger_task started");
+                loop {
+                    // NB: we purposefully don't do a select! for the cancellation here.
+                    // The cancellation would likely cause us to miss stderr messages.
+                    // We can rely on this to return from .await because when we SIGKILL
+                    // the child, the writing end of the stderr pipe gets closed.
+                    match stderr.readable_mut().await {
+                        Ok(mut guard) => {
+                            let mut errbuf: [u8; 16384] = [0; 16384];
+                            let res = guard.try_io(|fd| {
+                                use std::io::Read;
+                                fd.get_mut().read(&mut errbuf)
+                            });
+                            match res {
+                                Ok(Ok(0)) => {
+                                    // it closed the stderr pipe
+                                    break;
+                                }
+                                Ok(Ok(n)) => {
+                                    // The message might not be split correctly into lines here. But this is
+                                    // good enough, the important thing is to get the message to the log.
+                                    error!(
+                                        output=?String::from_utf8_lossy(&errbuf[0..n]),
+                                        "received output",
+                                    );
```

Comment:
The tests fail because the Debug print of `String` escapes the `"` in the postgres stderr.
I.e., postgres logs this line to stderr:

```
2023-10-17 14:35:35.400 GMT [283333] LOG:  registered custom resource manager "neon" with ID 134
```

and pageserver captures & logs it like so

```
2023-10-17T14:35:35.401721Z ERROR wal-redo-postgres-stderr{pid=283333 tenant_id=707e654029a6e4560c7d0f57effd8bc7 pg_version=16}: received output output="2023-10-17 14:35:35.400 GMT [283333] LOG:  registered custom resource manager \"neon\" with ID 134\n"
```

Before this PR, we would log something like

```
2023-10-18T06:10:31.925937Z ERROR ...:apply_wal_records{tenant_id=e0b9677876f0b99dd1fe053198a2388b pid=4128510}: wal-redo-postgres: 2023-10-18 06:10:31.925 GMT [4128510] LOG:  registered custom resource manager "neon" with ID 134
```

That's why the exemption for this in `neon_fixtures.py` doesn't catch it.


Now I'm wondering whats the right thing to do here:

1. wrong log level: This is not an ERROR but we're treating any walredo output on stderr as something ERROR-level.
   => make walredo not log this; walredo should stay quiet unless there's a problem  (will create an issue)
2. what to do about the scaping: I would want to preserve the structured logging that this PR introduces, but I guess I should switch to `Display` ?

Thoughts?